[🐶] Transform MCP Server Tools

strawgate opened the pull request:
1. Introduce serializable models for Tool and Argument Transform
2. Allow passing serializable Tool Transform models to FastMCP and ToolManager
3. Change MCPConfig to automatically load transforming MCP Servers
4. Allow using MCPConfigTransport without prefixes on all the servers

## Example Scenario
My dog likes to run, but he doesn't like to fetch, so let's change the fetch server to meet his needs:

## Yaml

```yaml
mcpServers:
    fetch:
        command: uvx
        args:
            - mcp-server-fetch
        tools:
            fetch:
                name: run_spot_run
                arguments:
                    url:
                        name: bark
                        description: bark bark bark
                        default: woof woof woof
```
`mcp_config = MCPConfig.from_dict(yaml.safe_load(yaml_config))`

## Dict

```python
    mcp_config = {
        "mcpServers": {
            "fetch": {
                "command": "uvx",
                "args": ["mcp-server-fetch"],
                "tools": {
                    "fetch": {
                        "name": "run_spot_run",
                        "arguments": {
                            "url": {
                                "name": "bark",
                                "description": "bark bark bark",
                                "default": "woof woof woof",
                            }
                        },
                    }
                },
            }
        }
    }
```

## Strongly Typed
```python
    mcp_config = MCPConfig(
        mcpServers={
            "stdio": TransformingStdioMCPServer(
                command="uvx",
                args=["mcp-server-fetch"],
                tools={
                    "fetch": ToolTransformConfig(
                        name="run_spot_run",
                        arguments={
                            "url": ArgTransformConfig(
                                name="bark",
                                description="bark bark bark",
                                default="woof woof woof",
                            )
                        },
                    )
                },
            )
        }
    )
```

Then load the config into a Proxy or Client:
```
proxy = FastMCP.as_proxy(mcp_config)
# Options for loading mcp_config described below
  
tools = await proxy.get_tools()
# Transformed tools appear
```

jlowin commented on src/fastmcp/server/proxy.py:
```diff
@@ -80,18 +93,11 @@ async def list_tools(self) -> list[Tool]:
-                    structured_content=result.structured_content,
-                )
+
+        if tool := await self.get_tool(key):
```
This is a little too reductive -- the `super().call_tool` method contains all error handling that tool.run() lacks.

jlowin reviewed (changes requested):
I really like this. I think we should skip contrib and go straight to main. I put minor review comments inline and two larger thoughts:

- I'm not crazy about `Request` as the class suffix as these aren't really inflight objects. I'd suggest `ArgTransformConfig` and `ToolTransformConfig`
- Some sort of flag to automatically disable any tools not included in the tools list. So that you could essentially do something like
```python
{
  my_server:
    cmd: whatever
    tools: {
        tool_1: True
        tool_2: True
        exclude_others: True # this is definitely not the right vocab or possibly position, just for illustration
    }
    exclude_tools: True # also definitely wrong vocab
}
```

and no matter what tools this thing had, tool_1 and tool_2 are the only two exposed (with no transforms). Not sure if True is the right sentinel and pretty sure `exclude_others` is just bad but I think it would level up this whole thing.

To that end, should this be a property of FastMCP rather than just FastMCPProxy? Like supplying tool transforms to a server could be useful without even knowing yet if you're transforming a proxy, a mounted nested subserver, or whatever. Maybe i'm getting ahead of ourselves on that and we start here. Just a thought.

strawgate commented on src/fastmcp/server/proxy.py:
```diff
@@ -425,6 +431,7 @@ def __init__(
         client: Client | None = None,
         *,
         client_factory: Callable[[], Client] | None = None,
+        tool_transforms: dict[str, ToolTransformRequest] | None = None,
```
Is it intentional that it's tools only and not just `transform`? Or?

strawgate commented on src/fastmcp/mcp_config.py:
```diff
@@ -195,9 +205,9 @@ def from_dict(cls, config: dict[str, Any]) -> MCPConfig:
 
             # Determine if this is stdio or remote based on fields
             if "command" in server_config:
-                parsed_servers[name] = StdioMCPServer.model_validate(server_config)
```
If we merge all of the functionality into MCPConfig then it can go away, but if we want to keep MCPConfig and have a separate class for including transformations then I think this allows them to share more logic between them.

Are you proposing just moving this functionality directly into MCPConfig?

strawgate replied:
merged it into mcpconfig

strawgate commented:
Okay, the latest version of the PR should address the not set field, moves the transformation logic into the regular Tool Manager, and updates MCPConfig to support transforming servers.

I also changed the to_dict/etc on MCPConfig and added some tests, @jlowin happy to undo those changes if needed.

Let me know what you think

jlowin commented on src/fastmcp/tools/tool_transform.py:
```diff
@@ -798,3 +829,43 @@ def _function_has_kwargs(fn: Callable[..., Any]) -> bool:
         )
+
+
+class ToolTransformConfig(FastMCPComponent):
```
I'm not sure if this should inherit from FastMCPComponent - that's the base class for e.g. Tools Resources etc and consolidates logic like enabling / disabling. It could ultimately include fields that aren't good to have on the ToolTransformCOnfig (e.g. fields we dont expose for transforming). I think it's better to write out the transforable fields explicitly here.

jlowin commented on src/fastmcp/tools/tool_transform.py:
```diff
@@ -798,3 +829,43 @@ def _function_has_kwargs(fn: Callable[..., Any]) -> bool:
+        )
+
+    @classmethod
+    def apply_to_tools(
```
This feels misplaced to me - its a classmethod but doesn't actually use the class at all? It could more simpler and more discoverable as a documented module function I think

jlowin commented on src/fastmcp/mcp_config.py:
```diff
@@ -162,85 +193,62 @@ def to_transport(self) -> StreamableHttpTransport | SSETransport:
+    | TransformingRemoteMCPServer
+)
+
+McpServersType = dict[str, MCPServerTypes]
```
Weak opinion that having both `MCPServerTypes` and `McpServersType` in this file is going to cause more difficult-to-debug typing issues than it prevents. If possible I'd recommend sticking with `dict[str, MCPServerTypes]` as its much clearer and will save users from the risk of accidentally misspelling the import if they actually need to reference this type in another file.

strawgate commented:
I also added the tags include/exclude:
```
    config = {
        "mcpServers": {
            "test_1": {
                "command": "python",
                "args": [str(script_path)],
                "tools": {
                    "add": {
                        "name": "transformed_add",
                        "tags": ["keep"],
                        "arguments": {
                            "a": {"name": "transformed_a"},
                            "b": {"name": "transformed_b"},
                        },
                    },
                },
                "include_tags": ["keep"]
            },
            "test_2": {
                "command": "python",
                "args": [str(script_path)],
            },
        }
    }
```

strawgate commented:
I also made MCPConfig into a cool Generic with ABC and BaseModel so it could share the full implementation, whether canonical or not, and then I undid that and just added a simple CanonicalMCPConfig along with tests.

jlowin commented:
![](https://media3.giphy.com/media/v1.Y2lkPTc5MGI3NjExaTVpdnhwODBjMGI2Y2VjZTVpY3JwcnVpNXd1djNtbHJmaW04MzF6biZlcD12MV9pbnRlcm5hbF9naWZfYnlfaWQmY3Q9Zw/j9pKCnYLbN1nqYcuHc/giphy.gif)

May make some adjustments to the "CanonicalMCPConfig" but this is great
